subsys/fs/ext2/ext2_impl.c

```diff
@@ -0,0 +1,1542 @@
+static char *strchrnul(const char *str, const char c)
+{
+	char *p = strchr(str, c);
+
+	if (!p) {
+		p = strchr(str, '\0');
+	}
+	return p;
+}
+
+/**
+ * @brief Find inode
+ *
+ * @note Inodes are 32 bit. When we return signed 64 bit number then we don't
+ *       lose any information.
+ *
+ * @param r_offset If not NULL then offset in directory of that entry is written here.
+ * @return Inode number or negative error code
+ */
+static int64_t find_dir_entry(struct ext2_inode *inode, const char *name, size_t len,
+		uint32_t *r_offset)
+{
+	int rc;
+	uint32_t block, block_off, offset = 0;
+	struct ext2_data *fs = inode->i_fs;
+
+	while (offset < inode->i_size) {
+		block = offset / fs->block_size;
+		block_off = offset % fs->block_size;
+
+		rc = ext2_fetch_inode_block(inode, block);
+		if (rc < 0) {
+			return rc;
+		}
+
+		struct ext2_disk_dentry *de =
+			(struct ext2_disk_dentry *)(inode_current_block_mem(inode) + block_off);
+		size_t maxlen = MAX(len, (size_t)de->de_name_len);
+
+		if (strncmp(de->de_name, name, maxlen) == 0) {
+
+			if (r_offset) {
+				/* Return offset*/
+				*r_offset = offset;
+			}
+
+			return (int64_t)de->de_inode;
+		}
+
+		/* move to next directory entry */
+		offset += de->de_rec_len;
+	}
+
+	return -EINVAL;
+}
+
+/* Inode operations --------------------------------------------------------- */
+
+ssize_t ext2_inode_read(struct ext2_inode *inode, void *buf, uint32_t offset, size_t nbytes)
+{
+	int rc = 0;
+	ssize_t read = 0;
+	uint32_t block_size = inode->i_fs->block_size;
+
+	while (read < nbytes && offset < inode->i_size) {
+
+		uint32_t block = offset / block_size;
+		uint32_t block_off = offset % block_size;
+
+		rc = ext2_fetch_inode_block(inode, block);
+		if (rc < 0) {
+			break;
+		}
+
+		uint32_t left_on_blk = block_size - block_off;
+		uint32_t left_in_file = inode->i_size - offset;
+		size_t to_read = MIN(nbytes, MIN(left_on_blk, left_in_file));
+
+		memcpy((uint8_t *)buf + read, inode_current_block_mem(inode) + block_off, to_read);
+
+		read += to_read;
+		offset += to_read;
+	}
+
+	if (rc < 0) {
+		return rc;
+	}
+	return read;
+}
+
+ssize_t ext2_inode_write(struct ext2_inode *inode, const void *buf, uint32_t offset, size_t nbytes)
+{
+	int rc = 0;
+	ssize_t written = 0;
+	uint32_t block_size = inode->i_fs->block_size;
+
+	while (written < nbytes) {
+		uint32_t block = offset / block_size;
+		uint32_t block_off = offset % block_size;
+
+		LOG_DBG("inode:%d Write to block %d (offset: %d-%zd/%d)",
+				inode->i_id, block, offset, offset + nbytes, inode->i_size);
+
+		rc = ext2_fetch_inode_block(inode, block);
+		if (rc < 0) {
+			break;
+		}
+
+		size_t to_write = MIN(nbytes, block_size - block_off);
+
+		memcpy(inode_current_block_mem(inode) + block_off, (uint8_t *)buf + written,
+				to_write);
+		LOG_DBG("Written %zd bytes at offset %d in block i%d", to_write, block_off, block);
+
+		rc = ext2_commit_inode_block(inode);
+		if (rc < 0) {
+			return rc;
```

Comment:
Seems like we can just break here too.
Anyway, shouldn't these be covered by LOG_DBG, just in case, as it will be hard to figure out which one has returned rc.